modules/homa-lite/src/lib.rs

```diff
@@ -117,90 +102,89 @@ pub mod module {
-	pub type TotalStakedAmount<T: Config> = StorageValue<_, Balance, ValueQuery>;
-
 	#[pallet::pallet]
 	pub struct Pallet<T>(_);
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
-		/// Request to mint some Liquid currency, by locking up the given amount of Staking
-		/// currency. The exchange does not happen immediately, but on when the batch is processed
-		/// The user then needs to manually claim the Liquid currency once it is ready.
+		/// Mint some Liquid currency, by locking up the given amount of Staking currency.
+		/// The exchange rate is calculated using the ratio of the total amount of the staking and
+		/// liquid currency. A portion is reducted (defined as T::MaxRewardPerEra) to make up for
+		/// the fact that staking is only effective from the next era on (on the relaychain).
 		///
 		/// Parameters:
 		/// - `amount`: The amount of Staking currency to be exchanged.
 		/// - `xcm_dest_weight`: The weight to be paid to the destination for the XCM transfer.
-		#[pallet::weight(< T as Config >::WeightInfo::request_mint())]
+		#[pallet::weight(< T as Config >::WeightInfo::mint())]
 		#[transactional]
-		pub fn request_mint(origin: OriginFor<T>, amount: Balance, xcm_dest_weight: Weight) -> DispatchResult {
+		pub fn mint(origin: OriginFor<T>, amount: Balance, xcm_dest_weight: Weight) -> DispatchResult {
 			let who = ensure_signed(origin)?;
-
-			// Ensure the amount is above the minimum
+			// Ensure the amount is above the minimum, after the MintFee is deducted.
 			ensure!(
-				amount >= T::MinimumMintThreshold::get(),
+				amount > T::MinimumMintThreshold::get() + T::MintFee::get(),
 				Error::<T>::MintAmountBelowMinimumThreshold
 			);
 
 			// Ensure the total amount staked doesn't exceed the cap.
-			let new_total_staked = Self::total_staked_amount()
+			let new_total_staked = Self::total_staking_currency()
 				.checked_add(amount)
 				.ok_or(ArithmeticError::Overflow)?;
 			ensure!(
 				new_total_staked <= Self::staking_currency_mint_cap(),
 				Error::<T>::ExceededStakingCurrencyMintCap
 			);
 
-			let current_batch = Self::current_batch();
 			let staking_currency = T::StakingCurrencyId::get();
 
 			// ensure the user has enough funds on their account.
 			T::Currency::ensure_can_withdraw(staking_currency, &who, amount)?;
 
-			// Cross-chain transfers the staking assets.
+			// Calculate how much Liquid currency is to be minted.
+			// Gets the current exchange rate
+			let staking_total = Self::total_staking_currency();
+			let liquid_total = T::Currency::total_issuance(T::LiquidCurrencyId::get());
+			let exchange_rate =
+				Ratio::checked_from_rational(liquid_total, staking_total).unwrap_or_else(T::DefaultExchangeRate::get);
+
+			// liquid_to_mint = ( (staked_amount - MintFee) * liquid_total / staked_total ) * (1 -
+			// MaxRewardPerEra)
+			let mut liquid_to_mint = exchange_rate
+				.checked_mul_int(
+					amount
+						.checked_sub(T::MintFee::get())
+						.expect("Arithmatic should not panic"),
```

Comment:
This is *not a proof*. It should like "Amount is ensured greater than T::MintFee; qed"